### ISSUE TITLE
E2E Tests: Review disabled aXe rules

### ISSUE THREAD

**Issue by swissspidy** [GoogleForCreators/web-stories-wp#7365]
## Task Description

When we initially added aXe testing to the e2e test setup, we've disable some rules that caused frequent errors, either because of known bugs in the editor or issues in WordPress.

https://github.com/google/web-stories-wp/blob/d0ffcff89e8840ad9fd0bbe7fd93d1e4235a635b/packages/e2e-tests/src/config/bootstrap.js#L204-L219

Let's go over each disabled rule to:

1. See whether the rule can be safely removed (i.e. there are no errors)
2. Identify where the issues are coming from (editor, dashboard, WordPress itself, something else) and their nature (valid bug, false positive)
3. Document the reason for it being here in this ticket
4. Create tickets to fix bugs where applicable

The goal is to reduce the number of disabled rules and improve the accessibility of the plugin.

**Comment by spacedmonkey** [comment]
All of these accessibility issues seem to be related to the editor. Many issues are relatively simple to fix. I have list of common issues / low hanging fruit to fix. I would recommend we create tickets for of these rules and re-enable them one by one. 

### aria-allowed-attr
```
    Rule: "aria-allowed-attr" (Elements must only use allowed ARIA attributes)
    Help: https://dequeuniversity.com/rules/axe/4.2/aria-allowed-attr?application=axe-puppeteer
    Affected Nodes:
      svg[aria-label="Media\ Gallery"]
        Fix ALL of the following:
          - ARIA attribute cannot be used: aria-label.
      svg[aria-label="Explore\ Media"]
        Fix ALL of the following:
          - ARIA attribute cannot be used: aria-label.
      #text-tab-icon
        Fix ALL of the following:
          - ARIA attribute cannot be used: aria-label.
      svg[aria-label="Shapes\ library"]
        Fix ALL of the following:
          - ARIA attribute cannot be used: aria-label.
      svg[aria-label="Page\ templates\ library"]
        Fix ALL of the following:
          - ARIA attribute cannot be used: aria-label.
      video
        Fix ALL of the following:
          - ARIA attribute cannot be used: aria-label.
      .frameElement__Wrapper-sc-1pyee9f-0
        Fix ALL of the following:
          - ARIA attribute cannot be used: aria-labelledby.
```

These errors seem to valid. For the most part there incorrect use of aria-label on svg icons.  Also it seems like video tags do not this attribute. 

### color contrast

```
    Rule: "color-contrast" (Elements must have sufficient color contrast)
    Help: https://dequeuniversity.com/rules/axe/4.2/color-contrast?application=axe-puppeteer
    Affected Nodes:
      #document > .tabview__TabText-sc-16p5pbf-2.dKBTHu.cBzzzP
        Fix ANY of the following:
        - Element has insufficient color contrast of 3.53 (foreground color: #767e80, background color: #26292a, font size: 10.5pt (14px), font weight: bold). Expected contrast ratio of 4.5:1
      #prepublish > .tabview__TabText-sc-16p5pbf-2.dKBTHu.cBzzzP
        Fix ANY of the following:
        - Element has insufficient color contrast of 3.53 (foreground color: #767e80, background color: #26292a, font size: 10.5pt (14px), font weight: bold). Expected contrast ratio of 4.5:1
```

These. color contrast issues seem simple enough to fix. We may need some feedback from UX. 

### duplicate-id
```
    Rule: "duplicate-id" (id attribute value must be unique)
    Help: https://dequeuniversity.com/rules/axe/4.2/duplicate-id?application=axe-puppeteer
    Affected Nodes:
      .metabox-location-normal > .sidebar-open
        Fix ANY of the following:
        - Document has multiple static elements with the same id attribute: poststuff
      .metabox-location-normal > .sidebar-open > .postbox-container
        Fix ANY of the following:
        - Document has multiple static elements with the same id attribute: postbox-container-2
```

Duplicate ids are generated in meta boxes. The id, is on a hidden form element, so changing this id, should have little to know effect. 

### aria-required-parent
```
    Rule: "aria-required-parent" (Certain ARIA roles must be contained by particular parents)
    Help: https://dequeuniversity.com/rules/axe/4.2/aria-required-parent?application=axe-puppeteer
    Affected Nodes:
      .pagepreview__Page-l9b0au-0
        Fix ANY of the following:
        - Required ARIA parent role not present: listbox
 ```

Parent aria attribute required in one place. 

### landmark-unique
```
 Rule: "landmark-unique" (Ensures landmarks are unique)
    Help: https://dequeuniversity.com/rules/axe/4.2/landmark-unique?application=axe-puppeteer
    Affected Nodes:
      .fmTQwD
        Fix ANY of the following:
        - The landmark must have a unique aria-label, aria-labelledby, or title to make landmarks distinguishable
```

```
<div aria-label="Fullbleed area" role="region" data-testid="fullbleed" class="layout__FullbleedContainer-sc-99rnh0-5 bAvKjO">
```

Seems like `aria-label` is repeated in document. 

### nested-interactive
```
    Rule: "nested-interactive" (Ensure interactive controls are not nested)
    Help: https://dequeuniversity.com/rules/axe/4.2/nested-interactive?application=axe-puppeteer
    Affected Nodes:
      #library-tab-text
        Fix ANY of the following:
        - Element has focusable descendants
      .carouselPage__ReorderablePage-sc-1ey6om8-3
        Fix ANY of the following:
        - Element has focusable descendants
```

<img width="80" alt="Screenshot 2021-05-13 at 14 54 25" src="https://user-images.githubusercontent.com/237508/118135605-242ca180-b3fb-11eb-8e3d-218b83ddf646.png">
This is basically a button nested inside a button. This is a problem 

<img width="88" alt="Screenshot 2021-05-13 at 14 56 23" src="https://user-images.githubusercontent.com/237508/118135870-6950d380-b3fb-11eb-8c4f-2069079d326f.png">
This is an button inside an option. 

### region
```
    Rule: "region" (All page content should be contained by landmarks)
    Help: https://dequeuniversity.com/rules/axe/4.2/region?application=axe-puppeteer
    Affected Nodes:
      #wpadminbar > .screen-reader-shortcut
        Fix ANY of the following:
        - Some page content is not contained by landmarks
``` 

This seems to WordPress related. There maybe a way to remove this element from the page or get it work correctly.

**Comment by spacedmonkey** [comment]
For reference, here is the PR with all the rules enabled. https://github.com/google/web-stories-wp/pull/7514

**Comment by spacedmonkey** [comment]
As this ticket is a discovery ticket, I am going move it review and assign @swissspidy. I have detailed the main issues and recommend some fixes. We should be able to create tickets for each of these and ticket them one by one.

**Comment by swissspidy** [comment]
> ### aria-allowed-attr
> ```
>     Rule: "aria-allowed-attr" (Elements must only use allowed ARIA attributes)
>     Help: https://dequeuniversity.com/rules/axe/4.2/aria-allowed-attr?application=axe-puppeteer
>     Affected Nodes:
>       svg[aria-label="Media\ Gallery"]
>         Fix ALL of the following:
>           - ARIA attribute cannot be used: aria-label.
>       svg[aria-label="Explore\ Media"]
>         Fix ALL of the following:
>           - ARIA attribute cannot be used: aria-label.
>       #text-tab-icon
>         Fix ALL of the following:
>           - ARIA attribute cannot be used: aria-label.
>       svg[aria-label="Shapes\ library"]
>         Fix ALL of the following:
>           - ARIA attribute cannot be used: aria-label.
>       svg[aria-label="Page\ templates\ library"]
>         Fix ALL of the following:
>           - ARIA attribute cannot be used: aria-label.
>       video
>         Fix ALL of the following:
>           - ARIA attribute cannot be used: aria-label.
>       .frameElement__Wrapper-sc-1pyee9f-0
>         Fix ALL of the following:
>           - ARIA attribute cannot be used: aria-labelledby.
> ```
> 
> These errors seem to valid. For the most part there incorrect use of aria-label on svg icons. Also it seems like video tags do not this attribute.

Tickets created:

* #7573
* #7574

> ### color contrast
> ```
>     Rule: "color-contrast" (Elements must have sufficient color contrast)
>     Help: https://dequeuniversity.com/rules/axe/4.2/color-contrast?application=axe-puppeteer
>     Affected Nodes:
>       #document > .tabview__TabText-sc-16p5pbf-2.dKBTHu.cBzzzP
>         Fix ANY of the following:
>         - Element has insufficient color contrast of 3.53 (foreground color: #767e80, background color: #26292a, font size: 10.5pt (14px), font weight: bold). Expected contrast ratio of 4.5:1
>       #prepublish > .tabview__TabText-sc-16p5pbf-2.dKBTHu.cBzzzP
>         Fix ANY of the following:
>         - Element has insufficient color contrast of 3.53 (foreground color: #767e80, background color: #26292a, font size: 10.5pt (14px), font weight: bold). Expected contrast ratio of 4.5:1
> ```
> 
> These. color contrast issues seem simple enough to fix. We may need some feedback from UX.

Ticket created:

* #7575
 
> ### duplicate-id
> ```
>     Rule: "duplicate-id" (id attribute value must be unique)
>     Help: https://dequeuniversity.com/rules/axe/4.2/duplicate-id?application=axe-puppeteer
>     Affected Nodes:
>       .metabox-location-normal > .sidebar-open
>         Fix ANY of the following:
>         - Document has multiple static elements with the same id attribute: poststuff
>       .metabox-location-normal > .sidebar-open > .postbox-container
>         Fix ANY of the following:
>         - Document has multiple static elements with the same id attribute: postbox-container-2
> ```
> 
> Duplicate ids are generated in meta boxes. The id, is on a hidden form element, so changing this id, should have little to know effect.

This ID is generated by WordPress and we don't really have control over it:

https://github.com/WordPress/wordpress-develop/blob/80302b7b35ede70c453c5cfcee00b963af878ea0/src/wp-admin/includes/post.php#L2265-L2279

Suggesting to keep disabled for now, with a comment explaining why.

> ### aria-required-parent
> ```
>     Rule: "aria-required-parent" (Certain ARIA roles must be contained by particular parents)
>     Help: https://dequeuniversity.com/rules/axe/4.2/aria-required-parent?application=axe-puppeteer
>     Affected Nodes:
>       .pagepreview__Page-l9b0au-0
>         Fix ANY of the following:
>         - Required ARIA parent role not present: listbox
> ```
> 
> Parent aria attribute required in one place.

Ticket created: #7576

> ### landmark-unique
> ```
>  Rule: "landmark-unique" (Ensures landmarks are unique)
>     Help: https://dequeuniversity.com/rules/axe/4.2/landmark-unique?application=axe-puppeteer
>     Affected Nodes:
>       .fmTQwD
>         Fix ANY of the following:
>         - The landmark must have a unique aria-label, aria-labelledby, or title to make landmarks distinguishable
> ```
> ```
> <div aria-label="Fullbleed area" role="region" data-testid="fullbleed" class="layout__FullbleedContainer-sc-99rnh0-5 bAvKjO">
> ```
> 
> Seems like `aria-label` is repeated in document.

Ticket created:

* #7578

> ### nested-interactive
> ```
>     Rule: "nested-interactive" (Ensure interactive controls are not nested)
>     Help: https://dequeuniversity.com/rules/axe/4.2/nested-interactive?application=axe-puppeteer
>     Affected Nodes:
>       #library-tab-text
>         Fix ANY of the following:
>         - Element has focusable descendants
>       .carouselPage__ReorderablePage-sc-1ey6om8-3
>         Fix ANY of the following:
>         - Element has focusable descendants
> ```
> 
> <img alt="Screenshot 2021-05-13 at 14 54 25" width="80" src="https://user-images.githubusercontent.com/237508/118135605-242ca180-b3fb-11eb-8e3d-218b83ddf646.png">
> 
> This is basically a button nested inside a button. This is a problem

I think this one has come up before in our audit or during initial implementation, and IIRC there wasn't really a better way to solve this. So we might need to keep this enabled in this case.

@barklund perhaps you can confirm.

> <img alt="Screenshot 2021-05-13 at 14 56 23" width="88" src="https://user-images.githubusercontent.com/237508/118135870-6950d380-b3fb-11eb-8c4f-2069079d326f.png">
> 
> This is an button inside an option.

This is covered by #7576 I think

> ### region
> ```
>     Rule: "region" (All page content should be contained by landmarks)
>     Help: https://dequeuniversity.com/rules/axe/4.2/region?application=axe-puppeteer
>     Affected Nodes:
>       #wpadminbar > .screen-reader-shortcut
>         Fix ANY of the following:
>         - Some page content is not contained by landmarks
> ```
> 
> This seems to WordPress related. There maybe a way to remove this element from the page or get it work correctly.

Yeah this is in the admin bar. Let's keep this one ignored for now. Don't want to change expected tab order and a11y tree for WP users outside of the editor.

**Comment by spacedmonkey** [comment]
Now that these tickets have been created, I feel like close this ticket out.

**Comment by swissspidy** [comment]
I wanted to get confirmation on the "Add text element" shortcut above, maybe it needs another ticket

> ### nested-interactive
> ```
>     Rule: "nested-interactive" (Ensure interactive controls are not nested)
>     Help: https://dequeuniversity.com/rules/axe/4.2/nested-interactive?application=axe-puppeteer
>     Affected Nodes:
>       #library-tab-text
>         Fix ANY of the following:
>         - Element has focusable descendants
>       .carouselPage__ReorderablePage-sc-1ey6om8-3
>         Fix ANY of the following:
>         - Element has focusable descendants
> ```
> 
> <img alt="Screenshot 2021-05-13 at 14 54 25" width="80" src="https://user-images.githubusercontent.com/237508/118135605-242ca180-b3fb-11eb-8e3d-218b83ddf646.png">
> 
> This is basically a button nested inside a button. This is a problem
>
> I think this one has come up before in our audit or during initial implementation, and IIRC there wasn't really a better way to solve this. So we might need to keep this enabled in this case.

@miina perhaps you are familiar with that shortcut?

**Comment by miina** [comment]
> @miina perhaps you are familiar with that shortcut?

Hm, not sure what could be a good alternative here since the tab itself is interactive and inside the tab there's a separate quick action that also has to be interactive.
A potential alternative could be removing the interactivity from the text tab and instead, have two interactive sibling buttons there. Might get a bit hacky to retain the same tab feel and looks, and keyboard interaction logic though. Thoughts?
